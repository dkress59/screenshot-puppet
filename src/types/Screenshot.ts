import { Request } from 'express'
import { PuppetOptions } from './PuppetOptions'

export class Screenshot {

	public w = 1024
	public h = 768
	public url = ''
	public src?: string
	public data?: Record<string, unknown>
	public fileName?: string
	public darkMode = false
	public remove?: string[]
	public errors: unknown[] = []
	public output: 'b64' | 'bin' | 'jpg' | 'json' | 'pdf' | 'png'  = 'json'

	constructor({ path, query }: Request, options?: PuppetOptions) {
		const { w, h, url, data, darkMode, remove, output } = query as Record<string, string>

		this.url = url.substring(0, 7) === 'http://'
			? url
			: 'https://' + url

		if (w) this.w = parseInt(w)
		if (h) this.h = parseInt(h)

		if (remove && typeof remove === 'string' && remove.split(',').length)
			this.remove = remove.split(',')

		const fileExt = path && path !== '/'
			? path.substring(1, path.length - 1).split('.').reverse()[0].toLowerCase()
			: false

		if (fileExt && ['jpg','jpeg','json','pdf','png'].indexOf(fileExt) > -1) {
			this.fileName = path.split('.').splice(-1,1).join('.')}

		if (options?.override === false) {
			if (options.output)
				this.output = options?.output
			if (options.data)
				this.data = options?.data
		} else {
			if (darkMode === 'true')
				this.darkMode = true // ToDo: reassure

			const mergedData = options?.data && data
				? {
					...options.data,
					...JSON.parse(data),
				}
				: data
					? JSON.parse(data)
					: options?.data

			if (mergedData)
				this.data = mergedData

			if (output && ['b64', 'bin', 'jpg', 'json', 'pdf', 'png'].indexOf(output) > -1)
				this.output = output as 'b64' | 'bin' | 'jpg' | 'json' | 'pdf' | 'png'
			else if (fileExt)
				this.output = fileExt === 'jpg'
					? 'jpg'
					: fileExt === 'jpeg'
						? 'jpg'
						: fileExt === 'json'
							? 'json'
							: fileExt === 'pdf'
								? 'pdf'
								: 'png'
		}

	}

}
