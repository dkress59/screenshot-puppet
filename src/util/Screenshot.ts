import queryString from 'query-string'
import { Request } from 'express'
import { ShotOptions } from '../Options'

interface ShotQuery {
	w?: number
	h?: number
	url?: string
	data?: string // JSON.stringify(Record<string, any>)
	dark?: boolean
	remove?: string[]
	output?: 'bin' | 'jpg' | 'json' | 'pdf' | 'png'
}

const mergeData = (options: ShotOptions | undefined, data: string | undefined): Record<string, unknown> | undefined => {
	const settings = options?.data
	const user = data

	if (settings && user)
		return {
			...settings,
			...JSON.parse(user) as Record<string, unknown>,
		}

	if (user)
		return JSON.parse(user) as Record<string, unknown>

	if (settings)
		return settings
}

export class Screenshot {

	public w = 1024
	public h = 768
	public url = ''
	public src: string | Buffer = ''
	public data?: Record<string, unknown>
	public fileName?: string
	public darkMode = false
	public remove?: string[]
	public errors: unknown[] = []
	public output: 'bin' | 'jpg' | 'json' | 'pdf' | 'png'  = 'json'

	constructor({ params, query }: Request, options?: ShotOptions) {
		const { w, h, url, data, dark, remove, output }: ShotQuery = queryString.parse(
			queryString.stringify(query as Record<string, string>), {
				arrayFormat: 'comma',
				parseBooleans: true,
				parseNumbers: true,
			}
		) 

		const formats = ['bin', 'jpg', 'json', 'pdf', 'png']

		if (url) // always true
			this.url = url.substring(0, 5) === 'http:'
				? url
				: url.substring(0, 6) === 'https:'
					? url
					: 'http://' + url

		if (remove?.length)
			this.remove = remove

		this.fileName = params && params.filename && params.filename.includes('.')
			? params.filename
				.split('.')
				.splice(0, params.filename.split('.').length - 1)
				.join('.')
			: undefined


		if (options?.override === false) {

			if (options.output)
				this.output = options.output
			if (options.data)
				this.data = options.data
			if (options.darkMode)
				this.darkMode = true

		} else {

			if (w) this.w = w
			if (h) this.h = h

			if (dark || options?.darkMode && !!dark)
				this.darkMode = true // ToDo: reassure

			if (mergeData(options, data))
				this.data = mergeData(options, data)

			const fileExt = params && params.filename
				? params.filename
					.split('.')
					.reverse()[0]
					.toLowerCase()
				: false

			if (options?.output && formats.includes(options.output))
				this.output = options.output
			if (fileExt)
				this.output = fileExt === 'jpeg'
					? 'jpg'
					: ['jpg', 'json', 'pdf', 'png'].includes(fileExt)
						? fileExt as 'jpg' | 'json' | 'pdf' | 'png'
						: 'png'
			if (output && formats.includes(output))
				this.output = output as 'bin' | 'jpg' | 'json' | 'pdf' | 'png'
		}

	}

}
