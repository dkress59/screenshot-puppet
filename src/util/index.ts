import util from 'util'
const redis = require('redis')
const REDIS = process.env.PUPPET_REDIS || 'redis://127.0.0.1:6379'

export const client = redis.createClient(REDIS)
client.get = util.promisify(client.get)
client.setex = util.promisify(client.setex)

import puppeteer, { Browser } from 'puppeteer'
import { Response } from 'express'

export const launchBrowser = async (res: Response): Promise<Browser> => {
	const browser = await puppeteer
		.launch({
			timeout: 666,
			//handleSIGINT: false,
			defaultViewport: null,
			args: ['--no-sandbox', '--disable-setuid-sandbox'],
		})
		.catch((e: any) => {
			res
				.status(500)
				.send({ error: 'error launching puppeteer: ' + e.toString() })
			throw new Error('error launching puppeteer: ' + e.toString())
		})

	return browser
}
