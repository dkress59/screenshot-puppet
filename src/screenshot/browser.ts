import { Response } from 'express'
import { Screenshot } from '../util/Screenshot'
import { logErrorToConsole, logToConsole } from '../util/utils'
import puppeteer, { Browser, LaunchOptions, PDFOptions, ScreenshotOptions } from 'puppeteer'


const qualities = new Set([
	0,
	1,
	2,
	3,
	4,
	5,
	6,
	7,
	8,
	9,
	10,
	11,
	12,
	13,
	14,
	15,
	16,
	17,
	18,
	19,
	20,
	21,
	22,
	23,
	24,
	25,
	26,
	27,
	28,
	29,
	30,
	31,
	32,
	33,
	34,
	35,
	36,
	37,
	38,
	39,
	40,
	41,
	42,
	43,
	44,
	45,
	46,
	47,
	48,
	49,
	50,
	51,
	52,
	53,
	54,
	55,
	56,
	57,
	58,
	59,
	60,
	61,
	62,
	63,
	64,
	65,
	66,
	67,
	68,
	69,
	70,
	71,
	72,
	73,
	74,
	75,
	76,
	77,
	78,
	79,
	80,
	81,
	82,
	83,
	84,
	85,
	86,
	87,
	88,
	89,
	90,
	91,
	92,
	93,
	94,
	95,
	96,
	97,
	98,
	99,
	100
])

export const launchBrowser = async (res?: Response, options?: LaunchOptions): Promise<Browser> => {
	process.setMaxListeners(16) // ToDo: options?
	return await puppeteer
		.launch({
			timeout: 6666,
			defaultViewport: null,
			ignoreHTTPSErrors: true,
			args: ['--no-sandbox', '--disable-setuid-sandbox'], // ToDo: neccessary?
			...options,
			headless: true,
		})
		.catch( /* istanbul ignore next */ (error) => {
			if (res)
				res
					.status(500)
					.send({
						message: 'error launching puppeteer',
						// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
						error: error.message,
					})

			throw new Error('error launching puppeteer: ' + JSON.stringify(error)) // breaks jest
		})
}


// eslint-disable-next-line sonarjs/cognitive-complexity
export const makeScreenshot = async (browser: Browser, image: Screenshot, options?: ScreenshotOptions | PDFOptions): Promise<Screenshot> => {

	const { w, h, url, darkMode, remove, output } = image

	const encoding = (output === 'bin')
		? 'binary'
		: 'base64'
	
	const type = options && ('type' in options)
		? options.type
		: output === 'pdf'
			? undefined
			: output === 'jpg'
				? 'jpeg'
				: 'png'

	const quality = (
		options
			&& 'quality' in options
			&& options.quality !== undefined
			&& qualities.has(options.quality)
	)
		? options.quality
		: undefined

	const safeOptions: ScreenshotOptions | PDFOptions = {
		...options,
		encoding,
		path: undefined, // can save to disk
		type,
		quality,
	}

	try {

		const page = await browser.newPage()
		await page.setViewport({ width: w, height: h })

		if (darkMode)
			await page.emulateMediaFeatures([{
				name: 'prefers-color-scheme',
				value: 'dark',
			}])

		await page.goto(url, {
			waitUntil: 'networkidle0'
		})

		if (remove)
			for (const sel of remove) {
				logToConsole('remove', sel)
				try {
					/* istanbul ignore next */
					await page.evaluate((sel) => {
						const nodes = document.querySelectorAll(sel)
						if (document.querySelectorAll(sel).length)
							for (let i = 0; i < nodes.length; i++)
								// eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
								nodes[i].remove()
					}, sel)
				} catch (error) {
					image.errors.push(error)
					logErrorToConsole(error)
				}
			}

		const screenshot = (output === 'pdf')
			? await page.pdf(safeOptions)
			: await page.screenshot(safeOptions)

		image.src = screenshot

	} catch (error) {

		logToConsole(error)
		image.errors.push(error)

	}


	return image

}
