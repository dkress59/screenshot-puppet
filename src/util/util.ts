import { Response } from 'express'
import { ICache } from '../types/Cache'
import ParsedQuery from '../types/ParsedQuery'
import Screenshot, { IQSC } from '../types/Screenshot'
import { client } from './browser'


export const logToConsole = (log1: unknown, log2?: unknown, log3?: unknown): void => {
	if (process.env.NODE_ENV !== 'dev') return

	console.log(log1)
	if (log2) console.log(log2)
	if (log3) console.log(log3)
	
}
export const logErrorToConsole = (log1: unknown, log2?: unknown, log3?: unknown): void => {
	if (process.env.NODE_ENV !== 'dev') return

	console.error(log1)
	if (log2) console.error(log2)
	if (log3) console.error(log3)
	
}


export const syncWithCache = async ({ image, cacheId, cached, needed, res } : {
	image: IQSC
	cacheId: string
	res?: Response
	cached?: ICache[]
	needed?: IQSC[]
}, method?: string): Promise<void> => {

	const src: string | null = await client.get(cacheId)
	const { link, title } = image

	if (method !== 'POST')
		try {
			if (!cached || !needed)
				throw new Error('cached[] and/or needed[] missing from syncWithCache.')

			if (src && src.length) {
				logToConsole('cached', cacheId)
				cached.push({
					src,
					link: link,
					title: title,
				})
			} else {
				logToConsole('needed', cacheId)
				needed.push(image as IQSC)
			}
		} catch (err) {
			logErrorToConsole(err)
			needed?.push(image as IQSC)
		}

	else
		try {
			if (!res)
				throw new Error('<Response> missing from syncWithCache.')

			if (src && src.length) {
				logToConsole('cached', cacheId)
				res.send(JSON.stringify({ src, link, title }))
				return
			} else {
				logToConsole('needed', cacheId)
			}
		} catch (err) {
			logErrorToConsole(err)
		}

}
