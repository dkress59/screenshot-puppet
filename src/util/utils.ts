import { Response } from 'express'
import { Cache } from '../types/Cache'
import { Screenshot } from '../types/Screenshot'
import { client } from './browser'


export const logToConsole = (log1: unknown, log2?: unknown, log3?: unknown): void | false => {
	if (process.env.NODE_ENV !== 'development') return
	if (!log1) return false

	log3
		? console.log(log1, log2, log3)
		: log2
			? console.log(log1, log2)
			: console.log(log1)
	
}
export const logErrorToConsole = (log1: unknown, log2?: unknown, log3?: unknown): void | false => {
	if (process.env.NODE_ENV !== 'development') return
	if (!log1) return false

	log3
		? console.error(log1, log2, log3)
		: log2
			? console.error(log1, log2)
			: console.error(log1)
	
}


export const syncWithCache = async ({ image, cacheId, cached, needed, res } : {
	image: Screenshot
	cacheId: string
	res?: Response
	cached?: Cache[]
	needed?: Screenshot[]
}, method?: string): Promise<void|true> => {

	const src: string | null = await client.get(cacheId)
	const { link, title } = image

	if (method !== 'POST')
		// GET
		try {
			if (!res)
				throw new Error('<Response> missing from syncWithCache.')

			if (src && src.length) {
				logToConsole('cached', cacheId)
				res.type('application/json')
				res.send(JSON.stringify({ src, link, title }))
				return true
			} else {
				logToConsole('needed', cacheId)
			}
		} catch (err) {
			logErrorToConsole(err)
		}

	else
		// POST
		try {
			if (!cached || !needed)
				throw new Error('cached[] and/or needed[] missing from syncWithCache.')

			if (src && src.length) {
				logToConsole('cached', cacheId)
				cached.push({
					src,
					link: link ?? '',
					title: title ?? '',
				})
			} else {
				logToConsole('needed', cacheId)
				needed.push(image)
			}
		} catch (err) {
			logErrorToConsole(err)
			needed?.push(image)
		}

}
