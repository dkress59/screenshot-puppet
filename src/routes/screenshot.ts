import { Response } from 'express'
import Screenshot from '../types/Screenshot'
import { logErrorToConsole, logToConsole } from '../util/utils'
import ParsedQuery from '../types/ParsedQuery'
import { launchBrowser, makeScreenshot } from '../util/browser'

export const postRouteScreenshot = async (req: Request, res: Response): Promise<void> => {

	res.type('application/json')
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	const { cached, needed } = req.body
	const browser = await launchBrowser(res)

	const returns = []
	const errors = []
	for await (const image of needed)
		returns.push(
			(async () => {
				try {

					const response = await makeScreenshot(browser, image)
					if (response.source)
						return response
					else
						errors.push(response)

				} catch (error) {

					console.error(error)
					errors.push({
						...image,
						error,
					})

				}
			})()
		)

	Promise.all(returns)
		.then((images) => {
			if (images.filter((i) => i && i.error !== undefined).length)
				res.status(200).send(JSON.stringify([...cached, ...images]))
		})
		.catch((err) => {
			logErrorToConsole(err)
			res.status(500).send(JSON.stringify({ error: err }))
		})
		.finally(() => {
			logToConsole('closing browser...')
			browser.close().catch((e: Error) => void e)
		})
}

export const getRouteScreenshot = async (req: Request, res: Response): Promise<void> => {
	res.type('application/json')
	
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	const image = new Screenshot(new ParsedQuery(req))
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	image.remove = req.query.remove.split(',')

	const browser = await launchBrowser(res)

	const response = await makeScreenshot(browser, image)
	response.source
		? res.status(200).send(JSON.stringify(response))
		: image.error
			? res.status(500).send(JSON.stringify(response))
			: res.status(500).send(
				JSON.stringify({
					...image,
					error: 'Error while retreiving screen shot.',
				})
			)

	logToConsole('closing browser...')
	await browser.close()
	return
}
