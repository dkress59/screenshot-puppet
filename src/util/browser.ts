import util from 'util'
import puppeteer, { Browser } from 'puppeteer'
import { Response } from 'express'
import Screenshot from '../types/Screenshot'
import { logErrorToConsole, logToConsole } from './utils'
// eslint-disable-next-line @typescript-eslint/no-var-requires
const redis = require('redis')
const REDIS = process.env.PUPPET_REDIS || 'redis://127.0.0.1:6379'


export const client = redis.createClient(REDIS)
client.get = util.promisify(client.get)
client.set = util.promisify(client.set)
client.setex = util.promisify(client.setex)
client.exists = util.promisify(client.exists)


export const launchBrowser = async (res?: Response, timeout?: number): Promise<Browser> => {
	process.setMaxListeners(16)
	const browser = await puppeteer
		.launch({
			timeout: timeout ? timeout : 6666,
			defaultViewport: null,
			ignoreHTTPSErrors: true,
			args: ['--no-sandbox', '--disable-setuid-sandbox'], // ToDo: neccessary?
		})
		.catch((e) => {
			if (res)
				res
					.status(500)
					.send({ error: 'error launching puppeteer: ' + e.toString() })
			throw new Error('error launching puppeteer: ' + e.toString())
		})

	return browser
}


export const makeScreenshot = async (browser: Browser, image: Screenshot): Promise<Screenshot> => {
	const { width, height, link, url, darkMode, remove } = image
	if (!link) image.error = 'No link provided for the screen shot.'
	if (!link) return image

	try {

		const page = await browser.newPage()
		if (width && height) await page.setViewport({ width, height })

		if (darkMode)
			await page.emulateMediaFeatures([{
				name: 'prefers-color-scheme',
				value: 'dark',
			}])

		await page.goto(url)

		if (remove)
			remove.map((sel: string) => {
				logToConsole('remove', sel)
				try {
					page.evaluate((sel) => {
						const nodes = document.querySelectorAll(sel)
						if (document.querySelectorAll(sel).length)
							for (let i = 0; i < nodes.length; i++)
								nodes[i].parentNode.removeChild(nodes[i])
					}, sel)
				} catch (error) {
					//image.error = error
					logErrorToConsole(error)
				}
				return
			})

		const screenshot = await page.screenshot()
		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
		// @ts-ignore
		image.source = screenshot.toString('base64') // ToDo: Why isn't btoa() working?

		const cacheId = `${link}-${width}x${height}`
		await client.setex(cacheId, 60 * 60 * 24 * 30, image.source)
		logToConsole(`cache set for ${cacheId}`)

	} catch (error) {

		logToConsole(error)
		image.error = error

	}
	return image
}

export const makePDF = async (doc: string): Promise<false | Buffer> => {
	const path = process.env.LV_REACT_APP_URL + '/invoice/' + doc
	const cookies = [{
		name: 'allowCookies',
		value: 'true',
		path: '/',
		expires: -1,
	}, {
		name: 'ageIsVerified',
		value: 'true',
		path: '/',
		expires: -1,
	}]

	const browser = await launchBrowser()
	const page = await browser.newPage()

	try {

		await page.goto(path, { waitUntil: 'domcontentloaded' })
		await page.setCookie(...cookies)
		await page.reload({ waitUntil: 'networkidle2' })
		await page.emulateMediaType('screen')// ToDo: 'print'

		const file = await page.pdf({ format: 'A4' })
		await client.setex(`lv-pdf/${doc}`, 60 * 60 * 24 * 30, file.toString('base64'))

		await browser.close()
		return file

	} catch (err) {

		logErrorToConsole(err)
		await browser.close()
		return false

	}
}
