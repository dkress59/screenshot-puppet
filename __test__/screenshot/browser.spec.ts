/**
 * @jest-environment jsdom
 */
import { Request, Response } from 'express'
import { launchBrowser, makeScreenshot } from '../../src/screenshot/browser'
import { Screenshot } from '../../src/util/Screenshot'

import { launch, Browser } from 'puppeteer'
import { mocked } from 'ts-jest/dist/utils/testing'
jest.mock('puppeteer')

const mockedLaunch = mocked(launch, true)

const env = process.env
const mockPage = {
	emulateMediaFeatures: () => null,
	evaluate: () => null,
	goto: () => null,
	screenshot: () => 'abc',
	setViewport: () => null,
	pdf: () => 'abc',
}
const mockBrowser = {
	close: () => null,
	newPage: () => (mockPage),
}

let mockRequest: Partial<Request> = {}
let mockResponse: Partial<Response> = {}


describe('Puppeteer Screenshot Mechanism', () => {
	beforeEach(() => {
		process.env.NODE_ENV = 'test'
		mockedLaunch.mockImplementation((): Promise<Browser> =>
			new Promise(resolve => resolve(mockBrowser as unknown as Browser))
		)
		mockRequest = {
			path: '/',
			query: {
				url: 'https://duckduckgo.com'
			}
		}
		mockResponse = {
			type: jest.fn().mockReturnThis(),
			status: jest.fn().mockReturnThis(),
			send: jest.fn().mockReturnThis(),
			end: jest.fn().mockReturnThis(),
		}
	})
	afterAll(() => { process.env = env })

	describe('launching browser', () => {
		it('returns browser', async() => {
			const browser = await launchBrowser()

			expect(mockResponse.status).not.toHaveBeenCalled()
			expect(mockResponse.send).not.toHaveBeenCalled()

			expect(browser).toEqual(mockBrowser)
		})

		/* it('fails and sends error', async() => {
			mockedLaunch.mockRejectedValue({ mock_error: 'mock error' })
			const browser = await launchBrowser()

			expect(mockResponse.status).toHaveBeenCalledWith(500)
			expect(mockResponse.send).toHaveBeenCalledWith({
				message: 'error launching puppeteer',
				error: 'mock error',
			})

			expect(await launchBrowser).toThrow()
		}) */
	})

	describe('retreiving screen shot', () => {

		it('resolves', async() => {
			/* // eslint-disable-next-line @typescript-eslint/ban-ts-comment
			// @ts-ignore
			jest.spyOn(document, 'querySelectorAll').mockImplementation(() => {
				const one = document.createElement('p')
				one.classList.add('one')
				const two = document.createElement('p')
				two.setAttribute('id', 'two')
				const nodeList = document.createElement('div')
					.appendChild(one)
					.appendChild(one)
					.appendChild(two)
				return nodeList
			}) */

			const mockOptions = {}
			const req: Partial<Request> = {
				path: '/filename.json',
				query: {
					w: '800',
					h: '600',
					url: 'start.duckduckgo.com',
					dark: '1',
					remove: JSON.stringify(['.one','#two']),
					output: 'png',
				},
			}

			const image = await makeScreenshot(mockBrowser as unknown as Browser, new Screenshot(req as Request), mockOptions)

			expect(image).toBeTruthy()
		})

		it('reject newPage', async() => {
			const mockOptions = {}
			const req: Partial<Request> = {
				path: '/filename.json',
				query: {
					w: '800',
					h: '600',
					url: 'start.duckduckgo.com',
					dark: '1',
					remove: JSON.stringify(['.one','#two']),
					output: 'png',
				},
			}

			const image = await makeScreenshot({
				...mockBrowser,
				newPage: () => ({
					...mockPage,
					evaluate: () => Promise.reject('rejection')
				})
			} as unknown as Browser, new Screenshot(req as Request), mockOptions)

			// + document.querySelector
			expect(image.errors).toEqual(['rejection', 'rejection'])
			expect(image).toMatchSnapshot()
		})

		it('reject evauluate', async() => {
			const mockOptions = {}

			const image = await makeScreenshot({
				...mockBrowser,
				newPage: () => Promise.reject('rejection')
			} as unknown as Browser, new Screenshot(mockRequest as Request), mockOptions)

			expect(image.errors).toEqual(['rejection'])
		})

	})

})