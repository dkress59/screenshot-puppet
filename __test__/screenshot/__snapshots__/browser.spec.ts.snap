// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Puppeteer Screenshot Mechanism retreiving screen shot reject newPage 1`] = `
Screenshot {
  "darkMode": true,
  "errors": Array [
    "rejection",
    "rejection",
  ],
  "fileName": undefined,
  "h": 600,
  "output": "png",
  "remove": Array [
    ".one",
    "#two",
  ],
  "src": "abc",
  "url": "http://start.duckduckgo.com",
  "w": 800,
}
`;
